upstream/upstream.go

```diff
@@ -654,11 +658,103 @@ func (u *Upstream) shouldSkip(req *common.NormalizedRequest) (reason error, skip
 		}
 	}
 
-	// TODO evm: if block can be determined from request and upstream is only full-node and block is historical skip
+	// if block can be determined from request and upstream is only full-node and block is historical skip
+	bn, ebn := req.EvmBlockNumber()
+	if ebn != nil || bn <= 0 {
+		return fmt.Errorf("failed to get valid request block number: %w", ebn), false
+	}
+
+	if u.config.Evm.NodeType == common.EvmNodeTypeFull {
+		if req.Network() == nil || req.Network().EvmStatePollerOf(u.Config().Id) == nil {
+			return fmt.Errorf("network or state poller is nil"), false
+		}
+		lb := req.Network().EvmStatePollerOf(u.Config().Id).LatestBlock()
+
+		if bn < lb-u.config.Evm.MaxAvailableRecentBlocks {
+			return common.NewErrEndpointMissingData(fmt.Errorf("block number %d is out of range (must be >= %d)", bn, lb-u.config.Evm.MaxAvailableRecentBlocks)), true
+		}
+
+	}
 
 	return nil, false
 }
 
+func (u *Upstream) detectNodeType() (common.EvmNodeType, error) {
+	// Step 1: Get the latest block number using the state poller
+	req := &common.NormalizedRequest{}
+	if req.Network() == nil || req.Network().EvmStatePollerOf(u.Config().Id) == nil {
+		return "", common.NewErrUpstreamClientInitialization(fmt.Errorf("network or state poller is nil"), u.config.Id)
+	}
+	lb := req.Network().EvmStatePollerOf(u.Config().Id).LatestBlock()
```

Comment:
We need to make sure by this point state poller is initialized and latest block is fetched at least once. If not done or straightforward we need to revise the related design.

Reply:
totally. as discussed we want to have a Provider/Endpoint layer to handle detecting features at that level probably. we will handle this functionality in an upcoming PR